### ISSUE TITLE
LoadError: cannot load such file -- bundler/gem_tasks, incorrect entry for Bundler in $LOAD_PATH

### ISSUE THREAD

**Issue by eregon** [rubygems/rubygems#3570]
This issue happens when using Bundler 2.1.4 on Ruby 2.3.8.
It doesn't happen on Ruby 2.4+, nor with Bundler 1 on Ruby 2.3.8.

See https://github.com/ruby/setup-ruby/issues/51

It can be reproduced easily with:
```
git clone https://github.com/anirbanmu/str_metrics.git 
cd str_metrics
git checkout 7e4645e3919ca85a7eac87146ba6797271bb3dbd

chruby 2.3.8
gem install bundler
bundle config --local path vendor/bundle
bundle install
bundle exec rake spec
```

That gives:
```
rake aborted!
LoadError: cannot load such file -- bundler/gem_tasks
/home/eregon/tmp/str_metrics/Rakefile:7:in `require'
/home/eregon/tmp/str_metrics/Rakefile:7:in `<top (required)>'
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/home/eregon/.rubies/ruby-2.3.8/bin/bundle:22:in `load'
/home/eregon/.rubies/ruby-2.3.8/bin/bundle:22:in `<main>'
(See full trace by running task with --trace)
zsh: exit 1     bundle exec rake spec
```

And the reason is that `$LOAD_PATH` is incorrect inside the Rakefile, adding `puts $LOAD_PATH` as the first line of the Rakefile gives:
```
/home/eregon/tmp/str_metrics/lib
/home/eregon/tmp/extensions/x86_64-linux/2.3.0-static/str_metrics-0.1.1
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-rspec-1.38.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-performance-1.5.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-0.81.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/unicode-display_width-1.7.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ruby-progressbar-1.10.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-3.9.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-mocks-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-expectations-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-support-3.9.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rexml-3.2.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rainbow-3.0.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/pry-0.13.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/parser-2.7.1.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/parallel-1.19.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/method_source-1.0.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/jaro_winkler-1.5.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/jaro_winkler-1.5.4
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ffi-1.12.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/ffi-1.12.2
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/diff-lcs-1.3/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/coderay-1.1.2/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib/gems/bundler-2.1.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ast-2.4.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0/x86_64-linux
rake aborted!
LoadError: cannot load such file -- bundler/gem_tasks
```

`/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib/gems/bundler-2.1.4/lib` does not exist.

If not using `bundle config --local path vendor/bundle` then everything works fine, because the `$LOAD_PATH` is then:
```
/home/eregon/tmp/str_metrics/lib
/home/eregon/tmp/extensions/x86_64-linux/2.3.0-static/str_metrics-0.1.1
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rubocop-rspec-1.38.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rubocop-performance-1.5.2/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rubocop-0.81.0/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/unicode-display_width-1.7.0/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/ruby-progressbar-1.10.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rspec-3.9.0/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rspec-mocks-3.9.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.9.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rspec-core-3.9.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rspec-support-3.9.2/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rexml-3.2.4/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rainbow-3.0.0/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/pry-0.13.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/parser-2.7.1.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/parallel-1.19.1/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/method_source-1.0.0/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/jaro_winkler-1.5.4/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/extensions/x86_64-linux/2.3.0-static/jaro_winkler-1.5.4
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/ffi-1.12.2/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/extensions/x86_64-linux/2.3.0-static/ffi-1.12.2
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/diff-lcs-1.3/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/coderay-1.1.2/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/ast-2.4.0/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/rake-12.3.3/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0/x86_64-linux
```

On Ruby 2.3 with Bundler 1 (and similar on Ruby 2.4 with Bundler 2) with `bundle config --local path vendor/bundle`, the `$LOAD_PATH` is instead:
```
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-1.17.3/lib
/home/eregon/tmp/str_metrics/lib
/home/eregon/tmp/extensions/x86_64-linux/2.3.0-static/str_metrics-0.1.1
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-rspec-1.38.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-performance-1.5.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-0.81.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/unicode-display_width-1.7.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ruby-progressbar-1.10.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-3.9.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-mocks-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-expectations-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-support-3.9.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rexml-3.2.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rainbow-3.0.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/pry-0.13.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/parser-2.7.1.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/parallel-1.19.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/method_source-1.0.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/jaro_winkler-1.5.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/jaro_winkler-1.5.4
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ffi-1.12.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/ffi-1.12.2
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/diff-lcs-1.3/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/coderay-1.1.2/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-1.17.3/lib/gems/bundler-1.17.3/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ast-2.4.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0/x86_64-linux
```
Which contains additionally `/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-1.17.3/lib` (correct) but also `/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-1.17.3/lib/gems/bundler-1.17.3/lib` (incorrect but probably harmless).

When looking at the ENV (relevant entries kept) for Bundler 2 on MRI 2.3.8 with `bundle config --local path vendor/bundle`, I see:
```
$ bundle exec rake spec
{"BUNDLER_ORIG_BUNDLER_VERSION"=>
  "BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL",
 "BUNDLER_ORIG_BUNDLE_BIN_PATH"=>
  "BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL",
 "BUNDLER_ORIG_BUNDLE_GEMFILE"=>
  "BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL",
 "BUNDLER_ORIG_GEM_HOME"=>"BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL",
 "BUNDLER_ORIG_GEM_PATH"=>"BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL",
 "BUNDLER_ORIG_MANPATH"=>":",
 "BUNDLER_ORIG_PATH"=>
  "/home/eregon/.rubies/ruby-2.3.8/bin:/home/eregon/bin:/home/eregon/code/truffleruby-ws/openjdk1.8.0_242-jvmci-20.1-b01/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/eregon/code/mx:/home/eregon/code/mspec/bin:/opt/crystal/bin",
 "BUNDLER_ORIG_RB_USER_INSTALL"=>
  "BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL",
 "BUNDLER_ORIG_RUBYLIB"=>"BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL",
 "BUNDLER_ORIG_RUBYOPT"=>"",
 "BUNDLER_VERSION"=>"2.1.4",
 "BUNDLE_BIN_PATH"=>
  "/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/exe/bundle",
 "BUNDLE_GEMFILE"=>"/home/eregon/tmp/str_metrics/Gemfile",
 "GEM_HOME"=>"/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0",
 "GEM_PATH"=>"",
 "HOME"=>"/home/eregon",
 "LANG"=>"en_US.UTF-8",
 "PATH"=>
  "/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/bin:/home/eregon/.rubies/ruby-2.3.8/bin:/home/eregon/bin:/home/eregon/code/truffleruby-ws/openjdk1.8.0_242-jvmci-20.1-b01/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/eregon/code/mx:/home/eregon/code/mspec/bin:/opt/crystal/bin",
 "PWD"=>"/home/eregon/tmp/str_metrics",
 "RUBYLIB"=>
  "/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib",
 "RUBYOPT"=>
  "-r/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib/bundler/setup ",
 "RUBY_ROOT"=>"/home/eregon/.rubies/ruby-2.3.8",
 "_"=>"/home/eregon/.rubies/ruby-2.3.8/bin/bundle"}

rake aborted!
LoadError: cannot load such file -- bundler/gem_tasks
/home/eregon/tmp/str_metrics/Rakefile:7:in `require'
/home/eregon/tmp/str_metrics/Rakefile:7:in `<top (required)>'
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/home/eregon/.rubies/ruby-2.3.8/bin/bundle:22:in `load'
/home/eregon/.rubies/ruby-2.3.8/bin/bundle:22:in `<main>'
(See full trace by running task with --trace)
```

So `RUBYLIB` is correct here, but somehow doesn't end up in the final `$LOAD_PATH`, so something must remove it and it's not `RUBYOPT=-r/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib/bundler/setup` itself:
```
$ RUBYLIB=/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib ruby -e 'puts $LOAD_PATH'
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0/x86_64-linux

$ RUBYLIB=/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib RUBYOPT=-r/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib/bundler/setup ruby -e 'puts $LOAD_PATH'
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib
/home/eregon/tmp/str_metrics/lib
/home/eregon/tmp/extensions/x86_64-linux/2.3.0-static/str_metrics-0.1.1
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-rspec-1.38.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-performance-1.5.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rubocop-0.81.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/unicode-display_width-1.7.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ruby-progressbar-1.10.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-3.9.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-mocks-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-expectations-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.9.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rspec-support-3.9.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rexml-3.2.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rainbow-3.0.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/pry-0.13.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/parser-2.7.1.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/parallel-1.19.1/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/method_source-1.0.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/jaro_winkler-1.5.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/jaro_winkler-1.5.4
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ffi-1.12.2/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/ffi-1.12.2
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/diff-lcs-1.3/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/coderay-1.1.2/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/gems/2.3.0/gems/bundler-2.1.4/lib/gems/bundler-2.1.4/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/ast-2.4.0/lib
/home/eregon/tmp/str_metrics/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/site_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby/2.3.0/x86_64-linux
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/vendor_ruby
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0
/home/eregon/.rubies/ruby-2.3.8/lib/ruby/2.3.0/x86_64-linux
```

Not really sure what's going on here, I would appreciate if you can debug it further.

Note that `gem update --system` solves this issue, but that is of course not a great workaround as my understanding is Bundler 2 is supposed to support Ruby 2.3 out of the box.

cc @anirbanmu @MSP-Greg

**Comment by eregon** [comment]
Not sure if this is a RubyGems or Bundler issue or both.

**Comment by deivid-rodriguez** [comment]
Thanks for the report @eregon!

I started having a look. For the moment I identified the culprit inside rubygems and considered potential workarounds that we can add to bundler, but I need to validate them.

Will keep you posted.

**Comment by eregon** [comment]
@deivid-rodriguez Any opinion on how to workaround this in https://github.com/ruby/setup-ruby/issues/51?

Maybe always use Bundler 1 on Ruby 2.3 (like already done on Ruby 2.2)?

Or upgrade RubyGems to a known version which fixes this (e.g. the version of RubyGems for 2.4)? (but then users might have a harder time to reproduce locally).

**Comment by deivid-rodriguez** [comment]
I'm currently fixing this, and plan to release the fix with bundler 2.2 very soon. I'm not specially interested in fixing ruby 2.3 issues since it's already End of Life'd, but the details of this issue got me interested because of the non-existing path that bundler is adding to the `$LOAD_PATH`.

Anyways, the issue got fixed in rubygems side with https://github.com/rubygems/rubygems/pull/1527, so upgrading to any rubygems version including that (2.6.2 and above) should get the issue fixed.

**Comment by deivid-rodriguez** [comment]
If you prefer to stick with bundler 1 on ruby 2.3 until bundler 2.2 is released for reproducibility, that sounds good to me :+1:.

**Comment by eregon** [comment]
> the details of this issue got me interested because of the non-existing path that bundler is adding to the `$LOAD_PATH`.

Yeah, that non-existing path happens on Ruby 2.3, 2.4 and 2.5.
On 2.6 that entry is just not there (but there is still the correct one in first position).

**Comment by eregon** [comment]
I think I'll wait for the Bundler 2.2 release then and check if it fixes this, it doesn't seem very urgent to need a workaround (besides `with: bundler: 1`) and since Bundler 2 officially supports 2.3 probably best to use it if possible.

**Comment by deivid-rodriguez** [comment]
Yeah, the explanation of why the issue is only hit on ruby 2.3 is that it doesn't include https://github.com/rubygems/rubygems/pull/1527, which starts activating the `bundler` gem inside the `bundler` binstub, and activating gems adds themselves to the `$LOAD_PATH`. That's why we get a correct entry in the front of the `$LOAD_PATH` on higher versions of rubygems.

**Comment by deivid-rodriguez** [comment]
> I think I'll wait for the Bundler 2.2 release then and check if it fixes this, it doesn't seem very urgent to need a workaround (besides with: bundler: 1) and since Bundler 2 officially supports 2.3 probably best to use it if possible.

That sounds good, I hope to add a test too, but 'll ping you on the PR so you can try it.

**Comment by eregon** [comment]
Actually the same error happened for Puma's CI on Ruby 2.3, although I didn't realize it was this bug back then: https://github.com/puma/puma/runs/553627456
```
$ bundle exec rake compile
rake aborted!
LoadError: cannot load such file -- bundler/setup
/Users/runner/runners/2.165.2/work/puma/puma/Rakefile:1:in `require'
/Users/runner/runners/2.165.2/work/puma/puma/Rakefile:1:in `<top (required)>'
/Users/runner/runners/2.165.2/work/puma/puma/.bundle/puma/ruby/2.3.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
/Users/runner/.rubies/ruby-2.3.8/bin/bundle:22:in `load'
/Users/runner/.rubies/ruby-2.3.8/bin/bundle:22:in `<main>'
(See full trace by running task with --trace)
```
(more recent run in https://github.com/eregon/puma/runs/647762853)

**Comment by deivid-rodriguez** [comment]
@eregon I believe #3626 should fix this.

**Comment by eregon** [comment]
@deivid-rodriguez Thanks, I confirm that fixes the issue on 2.3.8 by adding the line in `exe/bundler`.
Sorry I couldn't try it before.

**Comment by eregon** [comment]
@deivid-rodriguez Nice work, thank you for investigating and fixing this!

Could you ping when there is a release of Bundler including this fix?
I subscribed to this repository releases notifications, but not sure how that will work for Bundler releases.

**Comment by deivid-rodriguez** [comment]
No problem! Yes, I will ping you :+1:.

**Comment by eregon** [comment]
@deivid-rodriguez Do you know when would the next Bundler release go out?
I think I'll have to force Bundler 1 for Ruby 2.3 for `ruby/setup-ruby`, otherwise the `path` cannot be set (https://github.com/ruby/setup-ruby/issues/51) and there is no way to do caching in GitHub Actions (https://github.com/ruby/setup-ruby/issues/52).

**Comment by deivid-rodriguez** [comment]
We were waiting on some server side fixes, but we should be good to go now. No fixed date, but should happen soon.
